### PR TITLE
Introduce new optimization levels for BFloat16 training

### PR DESCRIPTION
**Pull request by rohithkrn**
Introduce new optimization levels `O4` and `O5` for bfloat16 mixed precision training.

### CODE REVIEW

**Comment by rohithkrn** [comment]
I haven't looked at the tests yet, but planning to look at them and modify them too. But since there's no CI anyway, would like to start the review process on this.

**Comment by lcskrishna** [comment]
@rohithkrn Make sure "run_amp" in L0 tests pass for this PR.

**Comment by rohithkrn** [comment]
Summary:

- Introduce new opt_levels `O4` and `O5`. Registry lists used in `O4` are same as `O1` in this PR and may need revisions in the future after more testing is done.

- Both `O4` and `O5` uses loss scale value of 1. Hence no changes made to the loss scaling APIs wrt `O4` and `O5` in this PR.

- Enabled tests in `L0\run_amp` excluding 
   - `test_rnn.py` (rnn cells are not enabled in pytorch with bfloat16), 
   - `test_fused_sgd` (gave up on it just by looking at the code, very difficult to understand what's going on there. Anyway, the test is failing with other optimization levels too.)
   - `test_multiple_models_optimizers_losses` (same reason as above, will enable in subsequent PRs) 

Here's the log of test results I ran on my local machine. [tests.log](https://github.com/ROCmSoftwarePlatform/apex/files/4637056/tests.log)
@lcskrishna @ashishfarmer @jeffdaily ready for review.

**Comment by sunway513** [comment]
@rohithkrn , are the FP16/BF16 unit tests passing with your PR?

**Comment by rohithkrn** [comment]
> @rohithkrn , are the FP16/BF16 unit tests passing with your PR?

@sunway513 Yes. I have attached the log of tests I ran in my previous comment for reference. Thnx.

**Comment by sunway513** [comment]
Thanks @rohithkrn!
